processor/batchprocessor/batch_processor.go

```diff
@@ -0,0 +1,223 @@
+type batchProcessor struct {
+	sender consumer.TraceConsumer
+	name   string
+	logger *zap.Logger
+
+	batch  *batch
+	ticker *ticker
+
+	sendBatchSize uint32
+	tickTime      time.Duration
+	timeout       time.Duration
+}
+
+var _ consumer.TraceConsumer = (*batchProcessor)(nil)
+
+// newBatchProcessor creates a new batcher that batches spans by node and resource
+func newBatchProcessor(params component.ProcessorCreateParams, sender consumer.TraceConsumer, cfg *Config) *batchProcessor {
+	bp := &batchProcessor{
+		name:   cfg.Name(),
+		sender: sender,
+		logger: params.Logger,
+
+		sendBatchSize: cfg.SendBatchSize,
+		tickTime:      cfg.TickTime,
+		timeout:       cfg.Timeout,
+	}
+
+	bp.batch = newBatch(bp)
+
+	// start ticker after options loaded in
+	bp.ticker = newTicker(bp)
+	go bp.ticker.start()
+
+	return bp
+}
+
+// ConsumeTrace implements batcher as a SpanProcessor and takes the provided spans and adds them to
+// batches
+func (bp *batchProcessor) ConsumeTrace(ctx context.Context, td data.TraceData) error {
+	bp.batch.add(td)
+	return nil
+}
+
+func (bp *batchProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{MutatesConsumedData: true}
+}
+
+// Start is invoked during service startup.
+func (bp *batchProcessor) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (bp *batchProcessor) Shutdown(context.Context) error {
+	// TODO: flush accumulated data.
+	return nil
+}
+
+func (bp *batchProcessor) sendItems(
+	itemsToProcess []data.TraceData,
+	resourceSpansCount uint32,
+	measure *stats.Int64Measure,
+) {
+	td := data.NewTraceData()
+	rss := td.ResourceSpans()
+
+	// TODO: Use resourceSpansCount to set capacity of the rss slice
+	// For now there is no way to do that with internal data structure API
+
+	for _, td := range itemsToProcess {
+		td.ResourceSpans().MoveTo(rss)
+	}
+
+	statsTags := []tag.Mutator{tag.Insert(processor.TagProcessorNameKey, bp.name)}
+	_ = stats.RecordWithTags(context.Background(), statsTags, measure.M(1))
+
+	// TODO: This process should be done in an async way, perhaps with a channel + goroutine worker(s)
+	_ = bp.sender.ConsumeTrace(context.Background(), td)
+}
+
+type batch struct {
+	parent                  *batchProcessor
+	mu                      sync.RWMutex
+	items                   []data.TraceData
+	totalResourceSpansCount uint32
+	totalSpanCount          uint32
+	lastSent                int64
+}
+
+func newBatch(parent *batchProcessor) *batch {
+	return &batch{
+		parent: parent,
+		items:  make([]data.TraceData, 0, initialBatchCapacity),
+	}
+}
+
+// add updates current batch by adding new TraceData object
+func (b *batch) add(td data.TraceData) {
+	newResourceSpansCount := td.ResourceSpans().Len()
+	if newResourceSpansCount == 0 {
+		return
+	}
+
+	var itemsToProcess []data.TraceData
+	var totalResourceSpansCount uint32
```

Comment:
The name of this var does not appear to reflect the purpose. Maybe `spanCountToProcess`?

Reply:
Yes, but it would be `ResourceSpansCountToProcess` then because it's counter for `ResourceSpans` objects, not for `Span`.